pkg/scheduler/sources/clusterschedulingpolicysnapshot/controller.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+// Package clusterschedulingpolicysnapshot features a controller that enqueues CRPs for the
+// scheduler to process where there is a change in their scheduling policy snapshots.
+package clusterschedulingpolicysnapshot
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	fleetv1beta1 "go.goms.io/fleet/apis/placement/v1beta1"
+	"go.goms.io/fleet/pkg/scheduler/queue"
+	"go.goms.io/fleet/pkg/utils/controller"
+)
+
+// Reconciler reconciles the change in scheduling policy snapshots.
+type Reconciler struct {
+	// client is the client the controller uses to access the hub cluster.
+	client client.Client
+	// schedulerWorkqueue is the workqueue in use by the scheduler.
+	schedulerWorkqueue queue.ClusterResourcePlacementSchedulingQueueWriter
+}
+
+// Reconcile reconciles the cluster scheduling policy snapshot.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	policySnapshotRef := klog.KRef("", req.Name)
+	startTime := time.Now()
+	klog.V(2).InfoS("Scheduler source reconciliation starts", "clusterSchedulingPolicySnapshot", policySnapshotRef)
+	defer func() {
+		latency := time.Since(startTime).Milliseconds()
+		klog.V(2).InfoS("Scheduler source reconciliation ends", "clusterSchedulingPolicySnapshot", policySnapshotRef, "latency", latency)
+	}()
+
+	// Retrieve the policy snapshot.
+	policySnapshot := &fleetv1beta1.ClusterSchedulingPolicySnapshot{}
+	if err := r.client.Get(ctx, req.NamespacedName, policySnapshot); err != nil {
+		klog.ErrorS(err, "Failed to get cluster scheduling policy snapshot", "clusterSchedulingPolicySnapshot", policySnapshotRef)
+		return ctrl.Result{}, controller.NewAPIServerError(true, client.IgnoreNotFound(err))
+	}
+
+	// Check if the policy snapshot has been deleted.
+	//
+	// Normally this would not happen as the event filter is set to filter out all deletion events.
+	if policySnapshot.DeletionTimestamp != nil {
+		// The policy snapshot has been deleted; ignore it.
+		return ctrl.Result{}, nil
+	}
+
+	// Verify if the policy snapshot is currently active.
+	isLatestVal, ok := policySnapshot.Labels[fleetv1beta1.IsLatestSnapshotLabel]
+	if !ok {
+		// The IsLatestSnapshot label is not present; normally this should never occur.
+		klog.ErrorS(controller.NewUnexpectedBehaviorError(fmt.Errorf("IsLatestSnapshotLabel is missing")),
```

Comment:
change to warning?
cause it's possible when the pod may exit after removing the active label before creating a new snapshot. When the new pod is created, CRP controller could be recovered by itself and the label will be corrected.

For the NewUnexpectedBehaviorError, it means it cannot be recovered by itself and needs oncall person to be involved.

Should we remove this controller.NewUnexpectedBehaviorError wrapper?

Reply:
The API mentioned that this label must be set for all policy snapshots; plus, for the deactivation part, this label should be set to `false` rather than removed I assume?

Reply:
you're right, the label should be always present but the value could be false.